### ISSUE TITLE
[Feature Request] metadata for rooms

### ISSUE THREAD

**Issue by lukasIO** [livekit/livekit-server#115]
Hi, 

when reading through the docs I came across this section in "Working with data":
> LiveKit participants have a metadata field that you could use to store application-specific data. For example, this could be the seat number around a table, or other state.

Maybe I misunderstood the use case here, but wouldn't something like a seat number rather be data that only needs to be updated for one entity - namely the room itself, instead for each of the participants?

I found I have the need for displaying state to the user that is room specific state, for example the current number of participants of other rooms. 
To update room specific state on each participant seems highly redundant so I was wondering if there is a possibility to have an api for changing a `metadata` property for the whole room?

Thanks!

**Comment by davidzhao** [comment]
@lukasIO the scenario is that each participant has a different seat number; and that could change if the participant moves to a different spot.

I can see wanting to store room specific state too. I think some of the other folks are also interested in this and have a discussion on our slack. We'd welcome contributions here.

**Comment by lukasIO** [comment]
> @lukasIO the scenario is that each participant has a different seat number; and that could change if the participant moves to a different spot.

That makes sense, I misunderstood.
Great, I'll try to help by contributing then!

**Comment by davidzhao** [comment]
@FeepsDev, I'm curious to learn about your use case for this feature. Do you mind sharing more with me?

1. what kind of data are you looking to store on the room?
2. how is this used by the particpants?
3. how often is the data updated? by whom?
4. approximately how large is the data?

Thanks!

**Comment by lukasIO** [comment]
Thank you for giving this some thought!
I'll answer to your [comment ](https://github.com/livekit/livekit-server/pull/118#issuecomment-922656003) here for visibility.

Background:
I'm in the process of porting our prototype of a 3D video call app to livekit from mediasoup.
It has a bit of a video-game feel to it so there are similar needs for updating game like state across the room.
For things like positional updates of each participant it's straight forward to let each participant publish their current positions via the existing apis, but, due to the interactive nature, there is also the possibility to change certain things that are bound to the whole room in realtime.

For example currently we are supporting streaming audio from a radio station which the participants can listen to.
To change the stream for everyone it would be great to be able to push updates to the room state in realtime.
Another example would be that we employ a concept of spaces in which multiple rooms are held. We want the participants to be able to see the current participant count of other rooms within the same space. This would probably have the highest update frequency and the update would probably be triggered by a webhook to `participant_joined` and `participant_left`.

In my case it would be sufficient to have this as a server api as we are looking to keep options for granular permissions that we would handle outside of livekit (e.g. in cloud functions/a backend).
The alternative, having something like a dedicated WS enabled backend to push realtime updates to clients, is an option, however I'd be thrilled if we can avoid that and thus are able to deploy also to serverless environments. (Plus we'd have to reeimplement a room logic, authentication etc. which livekit already handles).

To answer your questions directly:
1. Right now we think JSON objects would be enough for our usecase.
2. Our idea of potentially letting participants interact with it is through a secured endpoint which would then use the server-api based on our own granular permissions. The participants would need to be allowed to subscribe to room updates.
3. I think that certain things would be updated around one time per second and others a lot less often. In our use case it would be updated by a cloud function (triggered by webhook) through the server api. 
4. No idea as of now. It might of course get larger over time which would mean it might be more efficient to have more granular room update events like `roomUpdate(roomName, eventType, payload)`

**Comment by davidzhao** [comment]
thanks @lukasIO! that's helpful to know. Have you looked at [publishing data](https://docs.livekit.io/guides/room/data#data-message-from-participants) ?

this was originally designed to send data from a participant to another (i.e. chat), but a community member has contributed the capability of sending data packets via a server API.

with this, you will be able to deliver any payload to the current participants in the room.

**Comment by davidzhao** [comment]
agreed, I think the initial state is better solved by something like Room metadata, since we already send that down in the JoinResponse. Just wanted to see if sendData is something you could leverage since it already exists.

I'm happy to propose a design for Room metadata and share in this thread.

**Comment by FeepsDev** [comment]
@davidzhao I want to use Room metadata for application specific things. For exemple in my case I will use the metadata for storing "tags" to the rooms (Like MicroOnly, Room Language, NeedPassword, etc... ), so when a user browses the Room List, he can have more information about each room.

As far as I'm concerned, a participant can edit the Room metadata if he has a specific permission. The frequency at which the data will be modified is the same when a person changes a checkbox. The data isn't large, I may just serialize a Protocol Buffer Message in it.

**Comment by davidzhao** [comment]
So here's how we could build Room metadata support. First we have to make a few protocol updates

1. add a metadata field to the [Room](https://github.com/livekit/protocol/blob/main/livekit_models.proto#L6) protobuf
2. add a `UpdateRoomMetadata` method to [RoomService](https://github.com/livekit/protocol/blob/main/livekit_room.proto#L10), that could update metadata of the room
3. add a `RoomUpdate` message to [SignalResponse](https://github.com/livekit/protocol/blob/main/livekit_rtc.proto#L29)

then we could implement them on the server side, and then clients.
